File: src/executor/executor.ts
/**
 * Experiment Executor
 *
 * Orchestrates experiment execution across SUTs and cases.
 * Handles error isolation, progress reporting, and result collection.
 */

import { execSync } from "node:child_process";
import { arch, platform, version as nodeVersion } from "node:process";

import type { CaseDefinition, Primitive } from "../types/case.js";
import type { CorrectnessResult, EvaluationResult, Provenance } from "../types/result.js";
import type { SutDefinition } from "../types/sut.js";
import { MemoryMonitor, MemoryWarningLevel } from "./memory-monitor.js";
import { generateRunId } from "./run-id.js";

/**
 * Configuration for experiment execution.
 */
export interface ExecutorConfig {
	/** Continue execution if a single run fails */
	continueOnError: boolean;

	/** Number of repetitions per case (for statistical analysis) */
	repetitions: number;

	/** Random seed base (incremented per repetition) */
	seedBase: number;

	/** Timeout per run in milliseconds (0 = no timeout) */
	timeoutMs: number;

	/** Whether to collect provenance information */
	collectProvenance: boolean;

	/** Number of concurrent runs (1 = sequential) */
	concurrency?: number;

	/** Progress callback */
	onProgress?: (progress: ExecutionProgress) => void;

	/** Per-run callback (can be async) */
	onResult?: (result: EvaluationResult) => void | Promise<void>;

	/** Enable memory monitoring (default: true) */
	monitorMemory?: boolean;

	/** Memory warning threshold in MB (default: 1024) */
	memoryWarningThresholdMb?: number;

	/** Memory critical threshold in MB (default: 2048) */
	memoryCriticalThresholdMb?: number;

	/** Abort execution at critical memory level (default: false) */
	abortOnMemoryCritical?: boolean;
}

/**
 * Default executor configuration.
 */
export const DEFAULT_EXECUTOR_CONFIG: ExecutorConfig = {
	continueOnError: true,
	repetitions: 1,
	seedBase: 42,
	timeoutMs: 0,
	collectProvenance: true,
};

/**
 * Execution progress report.
 */
export interface ExecutionProgress {
	/** Total planned runs */
	total: number;

	/** Completed runs */
	completed: number;

	/** Failed runs */
	failed: number;

	/** Current SUT */
	currentSut?: string;

	/** Current case */
	currentCase?: string;

	/** Current repetition */
	currentRepetition?: number;

	/** Elapsed time in milliseconds */
	elapsedMs: number;
}

/**
 * Execution summary returned after completion.
 */
export interface ExecutionSummary {
	/** Total runs attempted */
	totalRuns: number;

	/** Successful runs */
	successfulRuns: number;

	/** Failed runs */
	failedRuns: number;

	/** Total elapsed time */
	elapsedMs: number;

	/** Results collected */
	results: EvaluationResult[];

	/** Errors encountered */
	errors: { runId: string; error: string }[];
}

/**
 * Execution plan for a single run.
 */
export interface PlannedRun {
	/** Deterministic run ID */
	runId: string;

	/** SUT to execute */
	sutId: string;

	/** Case to run against */
	caseId: string;

	/** Repetition number */
	repetition: number;

	/** Random seed */
	seed: number;

	/** Configuration overrides */
	config?: Record<string, unknown>;
}

/**
 * Get provenance information for reproducibility.
 * @param collectProvenance
 */
const getProvenance = (collectProvenance: boolean): Provenance => {
	if (!collectProvenance) {
		return {
			runtime: {
				platform: platform,
				arch: arch,
				nodeVersion: nodeVersion,
			},
		};
	}

	let gitCommit: string | undefined;
	let dirty = false;

	try {
		gitCommit = execSync("git rev-parse HEAD", { encoding: "utf-8", env: undefined }).trim();
		const status = execSync("git status --porcelain", { encoding: "utf-8", env: undefined });
		dirty = status.length > 0;
	} catch {
		// Git not available or not a git repo
	}

	return {
		runtime: {
			platform: platform,
			arch: arch,
			nodeVersion: nodeVersion,
		},
		gitCommit,
		dirty,
		timestamp: new Date().toISOString(),
	};
};

/**
 * Experiment executor.
 *
 * @template TInput - The resource type (e.g., Graph, Dataset)
 * @template TInputs - The algorithm inputs type
 * @template TResult - The algorithm result type
 */
export class Executor<TInput = unknown, TInputs = unknown, TResult = unknown> {
	private readonly config: ExecutorConfig;
	private readonly inputCache: Map<string, TInput>;
	private readonly memoryMonitor?: MemoryMonitor;

	constructor(config: Partial<ExecutorConfig> = {}) {
		this.config = {
			...DEFAULT_EXECUTOR_CONFIG,
			monitorMemory: true,
			...config,
		};
		this.inputCache = new Map();

		// Initialize memory monitor if enabled
		if (this.config.monitorMemory) {
			this.memoryMonitor = new MemoryMonitor({
				warningThresholdMb: this.config.memoryWarningThresholdMb,
				criticalThresholdMb: this.config.memoryCriticalThresholdMb,
				verbose: true,
				onWarningLevelChange: (level, stats) => {
					if (level === MemoryWarningLevel.CRITICAL && this.config.abortOnMemoryCritical) {
						throw new Error(
							`Memory critical threshold exceeded (${stats.rssMb.toFixed(1)}MB). Aborting execution.`,
						);
					}
				},
			});
		}
	}

	/**
	 * Plan execution runs without executing.
	 *
	 * @param suts - SUTs to execute
	 * @param cases - Cases to run against
	 * @returns Array of planned runs
	 */
	plan(
		suts: SutDefinition<TInputs, TResult>[],
		cases: CaseDefinition<TInput, TInputs>[],
	): PlannedRun[] {
		const runs: PlannedRun[] = [];

		for (const sutDef of suts) {
			for (const caseDef of cases) {
				for (let rep = 0; rep < this.config.repetitions; rep++) {
					const seed = this.config.seedBase + rep;
					const runId = generateRunId({
						sutId: sutDef.registration.id,
						caseId: caseDef.case.caseId,
						seed,
						repetition: rep,
					});

					runs.push({
						runId,
						sutId: sutDef.registration.id,
						caseId: caseDef.case.caseId,
						repetition: rep,
						seed,
					});
				}
			}
		}

		return runs;
	}

	/**
	 * Execute all planned runs.
	 *
	 * @param suts - SUTs to execute
	 * @param cases - Cases to run against
	 * @param metricsExtractor - Function to extract metrics from result
	 * @param plannedRuns - Optional pre-filtered planned runs (for parallel workers)
	 * @returns Execution summary with all results
	 */
	async execute(
		suts: SutDefinition<TInputs, TResult>[],
		cases: CaseDefinition<TInput, TInputs>[],
		metricsExtractor: (result: TResult) => Record<string, number>,
		plannedRuns?: PlannedRun[],
	): Promise<ExecutionSummary> {
		const startTime = performance.now();

		const effectivePlannedRuns = plannedRuns ?? this.plan(suts, cases);
		const sutMap = new Map(suts.map((s) => [s.registration.id, s]));
		const caseMap = new Map(cases.map((c) => [c.case.caseId, c]));

		// Use concurrency limit if specified, otherwise sequential
		const concurrency = this.config.concurrency ?? 1;

		if (concurrency <= 1) {
			// Sequential execution (original behavior)
			return this.executeSequential(
				effectivePlannedRuns,
				sutMap,
				caseMap,
				metricsExtractor,
				startTime,
			);
		}

		// Parallel execution with concurrency limit
		return this.executeParallel(
			effectivePlannedRuns,
			sutMap,
			caseMap,
			metricsExtractor,
			startTime,
			concurrency,
		);
	}

	/**
	 * Execute runs sequentially (original behavior).
	 * @param plannedRuns
	 * @param sutMap
	 * @param caseMap
	 * @param metricsExtractor
	 * @param startTime
	 * @internal
	 */
	private async executeSequential(
		plannedRuns: PlannedRun[],
		sutMap: Map<string, SutDefinition<TInputs, TResult>>,
		caseMap: Map<string, CaseDefinition<TInput, TInputs>>,
		metricsExtractor: (result: TResult) => Record<string, number>,
		startTime: number,
	): Promise<ExecutionSummary> {
		const results: EvaluationResult[] = [];
		const errors: { runId: string; error: string }[] = [];
		let completed = 0;
		let failed = 0;

		for (const run of plannedRuns) {
			const sutDef = sutMap.get(run.sutId);
			const caseDef = caseMap.get(run.caseId);

			if (!sutDef || !caseDef) {
				errors.push({ runId: run.runId, error: "SUT or case not found" });
				failed++;
				continue;
			}

			// Report progress
			if (this.config.onProgress) {
				this.config.onProgress({
					total: plannedRuns.length,
					completed,
					failed,
					currentSut: run.sutId,
					currentCase: run.caseId,
					currentRepetition: run.repetition,
					elapsedMs: performance.now() - startTime,
				});
			}

			try {
				const result = await this.executeRun(run, sutDef, caseDef, metricsExtractor);
				results.push(result);

				// Call onResult callback and await if it returns a promise
				if (this.config.onResult) {
					const callbackResult = this.config.onResult(result);
					if (callbackResult instanceof Promise) {
						await callbackResult;
					}
				}

				completed++;
			} catch (error) {
				const errorMessage = error instanceof Error ? error.message : String(error);
				errors.push({ runId: run.runId, error: errorMessage });
				failed++;

				if (!this.config.continueOnError) {
					throw error;
				}
			}
		}

		return {
			totalRuns: plannedRuns.length,
			successfulRuns: completed,
			failedRuns: failed,
			elapsedMs: performance.now() - startTime,
			results,
			errors,
		};
	}

	/**
	 * Execute runs in parallel with a concurrency limit.
	 * @param plannedRuns
	 * @param sutMap
	 * @param caseMap
	 * @param metricsExtractor
	 * @param startTime
	 * @param concurrency
	 * @internal
	 */
	private async executeParallel(
		plannedRuns: PlannedRun[],
		sutMap: Map<string, SutDefinition<TInputs, TResult>>,
		caseMap: Map<string, CaseDefinition<TInput, TInputs>>,
		metricsExtractor: (result: TResult) => Record<string, number>,
		startTime: number,
		concurrency: number,
	): Promise<ExecutionSummary> {
		const results: EvaluationResult[] = [];
		const errors: { runId: string; error: string }[] = [];

		// Thread-safe counters
		let completed = 0;
		let failed = 0;
		const mutex = { locked: false };
		const lockQueue: (() => void)[] = [];

		const acquireLock = () => {
			return new Promise<void>((resolve) => {
				if (mutex.locked) {
					lockQueue.push(resolve);
				} else {
					mutex.locked = true;
					resolve();
				}
			});
		};

		const releaseLock = () => {
			const next = lockQueue.shift();
			if (next) {
				next();
			} else {
				mutex.locked = false;
			}
		};

		// Process a single run
		const processRun = async (run: PlannedRun): Promise<void> => {
			const sutDef = sutMap.get(run.sutId);
			const caseDef = caseMap.get(run.caseId);

			if (!sutDef || !caseDef) {
				await acquireLock();
				errors.push({ runId: run.runId, error: "SUT or case not found" });
				failed++;
				releaseLock();
				return;
			}

			try {
				const result = await this.executeRun(run, sutDef, caseDef, metricsExtractor);

				await acquireLock();
				results.push(result);
				completed++;

				// Call onResult callback (checkpoint save)
				if (this.config.onResult) {
					await this.config.onResult(result);
				}

				// Report progress
				if (this.config.onProgress) {
					this.config.onProgress({
						total: plannedRuns.length,
						completed,
						failed,
						currentSut: run.sutId,
						currentCase: run.caseId,
						currentRepetition: run.repetition,
						elapsedMs: performance.now() - startTime,
					});
				}
				releaseLock();
			} catch (error) {
				await acquireLock();
				const errorMessage = error instanceof Error ? error.message : String(error);
				errors.push({ runId: run.runId, error: errorMessage });
				failed++;
				releaseLock();

				if (!this.config.continueOnError) {
					throw error;
				}
			}
		};

		// Worker pool: process runs in batches
		for (let index = 0; index < plannedRuns.length; index += concurrency) {
			const batch = plannedRuns.slice(index, index + concurrency);
			await Promise.all(batch.map(processRun));
		}

		return {
			totalRuns: plannedRuns.length,
			successfulRuns: completed,
			failedRuns: failed,
			elapsedMs: performance.now() - startTime,
			results,
			errors,
		};
	}

	/**
	 * Execute a single run.
	 * @param run
	 * @param sutDef
	 * @param caseDef
	 * @param metricsExtractor
	 */
	private async executeRun(
		run: PlannedRun,
		sutDef: SutDefinition<TInputs, TResult>,
		caseDef: CaseDefinition<TInput, TInputs>,
		metricsExtractor: (result: TResult) => Record<string, number>,
	): Promise<EvaluationResult> {
		const runStartTime = performance.now();
		let peakMemoryBytes = 0;

		// Check memory before execution
		if (this.memoryMonitor) {
			const initialLevel = this.memoryMonitor.check();
			if (initialLevel === MemoryWarningLevel.EMERGENCY) {
				throw new Error("Memory at emergency level before execution. Aborting.");
			}
		}

		// Load or reuse input resource for this case (cached to avoid reloading)
		const cacheKey = caseDef.case.caseId;
		let input = this.inputCache.get(cacheKey);
		if (!input) {
			// Apply timeout to resource loading (large graphs can hang during loading)
			input = await (this.config.timeoutMs > 0
				? Promise.race([
						caseDef.getInput(),
						new Promise<never>((_, reject) =>
							setTimeout(() => {
								reject(new Error(`Resource loading timeout after ${this.config.timeoutMs}ms`));
							}, this.config.timeoutMs),
						),
					])
				: caseDef.getInput());
			this.inputCache.set(cacheKey, input);

			// Check memory after resource loading
			if (this.memoryMonitor) {
				const stats = this.memoryMonitor.getStats();
				peakMemoryBytes = Math.max(peakMemoryBytes, stats.rssBytes);
			}
		}

		// Get algorithm inputs for this case
		const inputs = caseDef.getInputs();

		// Create SUT instance (factory now takes only config)
		const sut = sutDef.factory(run.config);

		// Execute with timeout if configured
		const sutResult = await (this.config.timeoutMs > 0
			? Promise.race([
					sut.run({ ...inputs, input }),
					new Promise<never>((_, reject) =>
						setTimeout(() => {
							reject(new Error(`Timeout after ${this.config.timeoutMs}ms`));
						}, this.config.timeoutMs),
					),
				])
			: sut.run({ ...inputs, input }));

		const executionTimeMs = performance.now() - runStartTime;

		// Check memory after execution
		let finalMemoryBytes = 0;
		if (this.memoryMonitor) {
			const stats = this.memoryMonitor.getStats();
			peakMemoryBytes = Math.max(peakMemoryBytes, stats.rssBytes);
			finalMemoryBytes = stats.rssBytes;
		}

		// Extract metrics
		const metrics = metricsExtractor(sutResult);

		// Build correctness result (basic - can be extended)
		const correctness: CorrectnessResult = {
			expectedExists: caseDef.case.expectedOutput !== undefined,
			producedOutput: true,
			valid: true, // Assume valid if no exception
			matchesExpected: null, // Would need comparison logic
		};

		// Build provenance
		const provenance = getProvenance(this.config.collectProvenance);
		provenance.executionTimeMs = executionTimeMs;

		// Add memory stats if monitoring is enabled
		if (this.memoryMonitor && this.config.collectProvenance) {
			provenance.peakMemoryBytes = peakMemoryBytes;
			provenance.finalMemoryBytes = finalMemoryBytes;
		}

		return {
			run: {
				runId: run.runId,
				sut: run.sutId,
				sutRole: sutDef.registration.role,
				sutVersion: sutDef.registration.version,
				caseId: run.caseId,
				caseClass: caseDef.case.caseClass,
				config: run.config as Record<string, Primitive> | undefined,
				seed: run.seed,
				repetition: run.repetition,
			},
			correctness,
			outputs: {
				summary: {
					// Could extract from sutResult
				},
			},
			metrics: {
				numeric: metrics,
			},
			provenance,
		};
	}
}

/**
 * Create a default executor with standard configuration.
 * @param config
 */
export const createExecutor = <TInput = unknown, TInputs = unknown, TResult = unknown>(
	config?: Partial<ExecutorConfig>,
): Executor<TInput, TInputs, TResult> => new Executor(config);

File: python/src/ppef/executor/executor.py
"""Experiment Executor.

Orchestrates experiment execution across SUTs and cases.
Handles error isolation, progress reporting, and result collection.

Triple loop order: SUT-major, Case-middle, Repetition-minor.
"""

from __future__ import annotations

import platform
import subprocess
import sys
import time
from collections.abc import Callable
from concurrent.futures import ProcessPoolExecutor
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any, Protocol

from ppef.executor.run_id import generate_run_id
from ppef.types.result import (
    CorrectnessResult,
    EvaluationResult,
    Provenance,
    ResultMetrics,
    ResultOutputs,
    RunContext,
)
from ppef.types.sut import SutRole

# ---------------------------------------------------------------------------
# Protocols for SUT / Case / MetricsExtractor
# ---------------------------------------------------------------------------


class SUT(Protocol):
    """Protocol for a System Under Test."""

    @property
    def id(self) -> str: ...

    @property
    def config(self) -> dict[str, Any]: ...

    def run(self, inputs: Any) -> Any: ...


class SutDefinition(Protocol):
    """Protocol for a SUT definition with registration metadata."""

    @property
    def registration(self) -> _SutRegistration: ...

    def factory(self, config: dict[str, Any] | None = None) -> SUT: ...


class _SutRegistration(Protocol):
    @property
    def id(self) -> str: ...

    @property
    def version(self) -> str: ...

    @property
    def role(self) -> SutRole: ...


class CaseDefinition(Protocol):
    """Protocol for a case definition."""

    @property
    def case(self) -> _CaseInfo: ...

    def get_input(self) -> Any: ...

    def get_inputs(self) -> Any: ...


class _CaseInfo(Protocol):
    @property
    def case_id(self) -> str: ...

    @property
    def case_class(self) -> str | None: ...

    @property
    def expected_output(self) -> dict[str, Any] | None: ...

    @property
    def version(self) -> str | None: ...


MetricsExtractor = Callable[[Any], dict[str, float]]


# ---------------------------------------------------------------------------
# Helper: convert arbitrary dict values to Primitive-safe values
# ---------------------------------------------------------------------------

Primitive = str | int | float | bool | None


def _to_primitive_record(
    config: dict[str, Any] | None,
) -> dict[str, Primitive] | None:
    if config is None:
        return None
    result: dict[str, Primitive] = {}
    for key, value in config.items():
        if value is None or isinstance(value, (str, int, float, bool)):
            result[key] = value
        else:
            import json

            result[key] = json.dumps(value)
    return result


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass
class ExecutorConfig:
    """Configuration for experiment execution."""

    continue_on_error: bool = True
    repetitions: int = 1
    seed_base: int = 42
    timeout_ms: int = 0
    collect_provenance: bool = True
    max_workers: int | None = None
    unsafe_in_process: bool = False
    on_progress: Callable[[ExecutionProgress], None] | None = None
    on_result: Callable[[EvaluationResult], None] | None = None
    monitor_memory: bool = True
    memory_warning_threshold_mb: int = 1024
    memory_critical_threshold_mb: int = 2048
    abort_on_memory_critical: bool = False
    base_dir: str | None = None


@dataclass
class ExecutionProgress:
    """Execution progress report."""

    total: int
    completed: int
    failed: int
    current_sut: str | None = None
    current_case: str | None = None
    current_repetition: int | None = None
    elapsed_ms: float = 0.0


@dataclass
class ExecutionSummary:
    """Execution summary returned after completion."""

    total_runs: int
    successful_runs: int
    failed_runs: int
    elapsed_ms: float
    results: list[EvaluationResult] = field(default_factory=list)
    errors: list[dict[str, str]] = field(default_factory=list)


@dataclass
class PlannedRun:
    """Execution plan for a single run."""

    run_id: str
    sut_id: str
    case_id: str
    repetition: int
    seed: int
    config: dict[str, Any] | None = None


# ---------------------------------------------------------------------------
# Provenance collection
# ---------------------------------------------------------------------------


def _get_provenance(collect_provenance: bool) -> Provenance:
    """Collect provenance information for reproducibility."""
    runtime = {
        "platform": sys.platform,
        "arch": platform.machine(),
        "pythonVersion": platform.python_version(),
    }

    if not collect_provenance:
        return Provenance(runtime=runtime)

    git_commit: str | None = None
    dirty: bool | None = None

    try:
        git_commit = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            capture_output=True,
            text=True,
            check=True,
        ).stdout.strip()
        status_output = subprocess.run(
            ["git", "status", "--porcelain"],
            capture_output=True,
            text=True,
            check=True,
        ).stdout
        dirty = len(status_output) > 0
    except (subprocess.CalledProcessError, FileNotFoundError):
        pass

    return Provenance(
        runtime=runtime,
        git_commit=git_commit,
        dirty=dirty,
        timestamp=datetime.now(UTC).isoformat(),
    )


# ---------------------------------------------------------------------------
# Memory monitoring helper
# ---------------------------------------------------------------------------


def _get_memory_bytes() -> int:
    """Get current RSS memory usage in bytes.

    Uses resource.getrusage on Unix, falls back to 0 on unsupported platforms.
    """
    try:
        import resource

        usage = resource.getrusage(resource.RUSAGE_SELF)
        # On macOS, ru_maxrss is in bytes; on Linux, it is in kilobytes.
        if sys.platform == "darwin":
            return usage.ru_maxrss
        return usage.ru_maxrss * 1024
    except (ImportError, AttributeError):
        return 0


# ---------------------------------------------------------------------------
# Executor
# ---------------------------------------------------------------------------


class Executor:
    """Experiment executor.

    Orchestrates experiment execution across SUTs and cases with
    configurable concurrency, progress reporting, and error isolation.
    """

    def __init__(self, config: ExecutorConfig | None = None) -> None:
        self._config = config or ExecutorConfig()
        self._input_cache: dict[str, Any] = {}

    @property
    def config(self) -> ExecutorConfig:
        return self._config

    # ------------------------------------------------------------------
    # Plan
    # ------------------------------------------------------------------

    def plan(
        self,
        suts: list[SutDefinition],
        cases: list[CaseDefinition],
    ) -> list[PlannedRun]:
        """Generate execution plan without running.

        Loop order: SUT-major, Case-middle, Repetition-minor.
        Seed derivation: seed = seed_base + repetition_index.
        """
        runs: list[PlannedRun] = []

        for sut_def in suts:
            for case_def in cases:
                for rep in range(self._config.repetitions):
                    seed = self._config.seed_base + rep
                    run_id = generate_run_id(
                        {
                            "sutId": sut_def.registration.id,
                            "caseId": case_def.case.case_id,
                            "seed": seed,
                            "repetition": rep,
                        }
                    )
                    runs.append(
                        PlannedRun(
                            run_id=run_id,
                            sut_id=sut_def.registration.id,
                            case_id=case_def.case.case_id,
                            repetition=rep,
                            seed=seed,
                        )
                    )

        return runs

    # ------------------------------------------------------------------
    # Execute
    # ------------------------------------------------------------------

    def execute(
        self,
        suts: list[SutDefinition],
        cases: list[CaseDefinition],
        metrics_extractor: MetricsExtractor,
        planned_runs: list[PlannedRun] | None = None,
    ) -> ExecutionSummary:
        """Execute all planned runs.

        When ``unsafe_in_process`` is True, runs execute in the current
        process (useful for debugging).  Otherwise, runs are dispatched
        to a ``ProcessPoolExecutor`` for worker isolation.
        """
        start = time.perf_counter()
        effective_runs = planned_runs if planned_runs is not None else self.plan(suts, cases)

        sut_map = {s.registration.id: s for s in suts}
        case_map = {c.case.case_id: c for c in cases}

        if self._config.unsafe_in_process:
            return self._execute_in_process(
                effective_runs, sut_map, case_map, metrics_extractor, start
            )

        max_workers = self._config.max_workers or 1
        if max_workers <= 1:
            # Sequential but still conceptually "isolated" -- for true
            # process isolation we would need picklable SUTs.  Fall back
            # to in-process sequential execution here.
            return self._execute_in_process(
                effective_runs, sut_map, case_map, metrics_extractor, start
            )

        return self._execute_parallel(
            effective_runs, sut_map, case_map, metrics_extractor, start, max_workers
        )

    # ------------------------------------------------------------------
    # Sequential (in-process) execution
    # ------------------------------------------------------------------

    def _execute_in_process(
        self,
        planned_runs: list[PlannedRun],
        sut_map: dict[str, SutDefinition],
        case_map: dict[str, CaseDefinition],
        metrics_extractor: MetricsExtractor,
        start: float,
    ) -> ExecutionSummary:
        results: list[EvaluationResult] = []
        errors: list[dict[str, str]] = []
        completed = 0
        failed = 0

        for run in planned_runs:
            sut_def = sut_map.get(run.sut_id)
            case_def = case_map.get(run.case_id)

            if sut_def is None or case_def is None:
                errors.append({"runId": run.run_id, "error": "SUT or case not found"})
                failed += 1
                continue

            try:
                result = self._execute_run(run, sut_def, case_def, metrics_extractor)
                results.append(result)

                if self._config.on_result is not None:
                    self._config.on_result(result)

                completed += 1

                if self._config.on_progress is not None:
                    self._config.on_progress(
                        ExecutionProgress(
                            total=len(planned_runs),
                            completed=completed,
                            failed=failed,
                            current_sut=run.sut_id,
                            current_case=run.case_id,
                            current_repetition=run.repetition,
                            elapsed_ms=(time.perf_counter() - start) * 1000,
                        )
                    )
            except Exception as exc:
                errors.append({"runId": run.run_id, "error": str(exc)})
                failed += 1

                if not self._config.continue_on_error:
                    raise

        return ExecutionSummary(
            total_runs=len(planned_runs),
            successful_runs=completed,
            failed_runs=failed,
            elapsed_ms=(time.perf_counter() - start) * 1000,
            results=results,
            errors=errors,
        )

    # ------------------------------------------------------------------
    # Parallel execution (ProcessPoolExecutor)
    # ------------------------------------------------------------------

    def _execute_parallel(
        self,
        planned_runs: list[PlannedRun],
        sut_map: dict[str, SutDefinition],
        case_map: dict[str, CaseDefinition],
        metrics_extractor: MetricsExtractor,
        start: float,
        max_workers: int,
    ) -> ExecutionSummary:
        """Execute runs in parallel batches using ProcessPoolExecutor.

        Note: SUTs and cases must be picklable for cross-process transfer.
        If they are not, fall back to ``unsafe_in_process=True``.
        """
        results: list[EvaluationResult] = []
        errors: list[dict[str, str]] = []
        completed = 0
        failed = 0

        with ProcessPoolExecutor(max_workers=max_workers) as pool:
            futures = {}
            for run in planned_runs:
                sut_def = sut_map.get(run.sut_id)
                case_def = case_map.get(run.case_id)

                if sut_def is None or case_def is None:
                    errors.append({"runId": run.run_id, "error": "SUT or case not found"})
                    failed += 1
                    continue

                future = pool.submit(self._execute_run, run, sut_def, case_def, metrics_extractor)
                futures[future] = run

            for future in futures:
                run = futures[future]
                try:
                    result = future.result()
                    results.append(result)

                    if self._config.on_result is not None:
                        self._config.on_result(result)

                    completed += 1

                    if self._config.on_progress is not None:
                        self._config.on_progress(
                            ExecutionProgress(
                                total=len(planned_runs),
                                completed=completed,
                                failed=failed,
                                current_sut=run.sut_id,
                                current_case=run.case_id,
                                current_repetition=run.repetition,
                                elapsed_ms=(time.perf_counter() - start) * 1000,
                            )
                        )
                except Exception as exc:
                    errors.append({"runId": run.run_id, "error": str(exc)})
                    failed += 1

                    if not self._config.continue_on_error:
                        raise

        return ExecutionSummary(
            total_runs=len(planned_runs),
            successful_runs=completed,
            failed_runs=failed,
            elapsed_ms=(time.perf_counter() - start) * 1000,
            results=results,
            errors=errors,
        )

    # ------------------------------------------------------------------
    # Single run execution
    # ------------------------------------------------------------------

    def _execute_run(
        self,
        run: PlannedRun,
        sut_def: SutDefinition,
        case_def: CaseDefinition,
        metrics_extractor: MetricsExtractor,
    ) -> EvaluationResult:
        """Execute a single planned run and return an EvaluationResult."""
        run_start = time.perf_counter()
        peak_memory_bytes = 0

        # Memory check before execution
        if self._config.monitor_memory:
            mem = _get_memory_bytes()
            peak_memory_bytes = max(peak_memory_bytes, mem)
            if (
                self._config.abort_on_memory_critical
                and mem > self._config.memory_critical_threshold_mb * 1024 * 1024
            ):
                raise MemoryError(
                    f"Memory critical threshold exceeded ({mem / (1024 * 1024):.1f}MB). "
                    "Aborting execution."
                )

        # Load or reuse input resource (cached)
        cache_key = case_def.case.case_id
        if cache_key not in self._input_cache:
            self._input_cache[cache_key] = case_def.get_input()

            if self._config.monitor_memory:
                peak_memory_bytes = max(peak_memory_bytes, _get_memory_bytes())

        resource_input = self._input_cache[cache_key]

        # Get algorithm inputs
        inputs = case_def.get_inputs()

        # Create SUT instance
        sut = sut_def.factory(run.config)

        # Combine inputs with loaded resource
        run_inputs = {**inputs, "input": resource_input} if isinstance(inputs, dict) else inputs

        # Execute SUT
        sut_result = sut.run(run_inputs)

        execution_time_ms = (time.perf_counter() - run_start) * 1000

        # Memory after execution
        final_memory_bytes = 0
        if self._config.monitor_memory:
            final_memory_bytes = _get_memory_bytes()
            peak_memory_bytes = max(peak_memory_bytes, final_memory_bytes)

        # Extract metrics
        metrics = metrics_extractor(sut_result)

        # Build correctness
        correctness = CorrectnessResult(
            expected_exists=case_def.case.expected_output is not None,
            produced_output=True,
            valid=True,
            matches_expected=None,
        )

        # Build provenance
        provenance = _get_provenance(self._config.collect_provenance)
        provenance.execution_time_ms = execution_time_ms

        if self._config.monitor_memory and self._config.collect_provenance:
            provenance.peak_memory_bytes = peak_memory_bytes
            provenance.final_memory_bytes = final_memory_bytes

        return EvaluationResult(
            run=RunContext(
                run_id=run.run_id,
                sut=run.sut_id,
                sut_role=sut_def.registration.role,
                sut_version=sut_def.registration.version,
                case_id=run.case_id,
                case_class=case_def.case.case_class,
                config=_to_primitive_record(run.config),
                seed=run.seed,
                repetition=run.repetition,
            ),
            correctness=correctness,
            outputs=ResultOutputs(summary={}),
            metrics=ResultMetrics(numeric=metrics),
            provenance=provenance,
        )


def create_executor(config: ExecutorConfig | None = None) -> Executor:
    """Create a default executor with standard configuration."""
    return Executor(config)
